Added WinCNG support for AES CTR mode

madwizard-thomas opened the pull request:
I have added support for AES CTR mode (aes128-ctr, aes192-ctr, aes256-ctr) in the WinCNG backend. It is based on ECB mode similar to the OpenSSL implementation. With these modes supported the WinCNG backend will also be compatible again with the latest OpenSSH server settings so the test suite will run with this backand now as well.

mback2k commented:
Thanks. Do you think it might be possible to abstract the CTR-on-ECB-mode implementation and make the OpenSSL and WinCNG backends use the same abstraction layer? I would prefer to have just one CTR-on-ECB-mode implementation in libssh2. What do you think?

madwizard-thomas commented:
That might be a good idea, I'll see if I can look into that. I think libgcrypt has a native ctr implementation so it wouldn't be shared across all backends.
I've also looked into GCM mode but I think the abstract crypto layer would need some more changes to support additional authenticated data.

madwizard-thomas commented:
Hmm looking a bit further into it there's only a very small part of the code that is similar. The crypto functionality works very differently in the backends and I think it would be hard to make an abstraction and still fit the cipher type/context model that is used. The only code that would be shared is the xor and increment of the counter:
```c
 for (i = 0; i < 16; i++)
        *out++ = *in++ ^ b1[i];

    i = 15;
    while (c->ctr[i]++ == 0xFF) {
        if (i == 0)
            break;
        i--;
    }
```
And actually the first loop is different because in the WinCNG backend the xor happens in place.

mback2k commented on src/wincng.c:
```diff
@@ -253,6 +253,17 @@ _libssh2_wincng_init(void)
+        ret = BCryptSetProperty(_libssh2_wincng.hAlgAES_ECB, BCRYPT_CHAINING_MODE,
+            (PBYTE)BCRYPT_CHAIN_MODE_ECB,
+            sizeof(BCRYPT_CHAIN_MODE_ECB), 0);
+        if (!BCRYPT_SUCCESS(ret)) {
```
Please align the lines with the opening ( , just like in lines 249 and 250. Otherwise this patch looks really good. I am looking forward to testing and trying it soon.

mback2k commented on src/wincng.c:
```diff
@@ -1707,67 +1718,114 @@ _libssh2_wincng_cipher_init(_libssh2_cipher_ctx *ctx,
+    pbIV = NULL;
+    pbCtr = NULL;
+    dwIV = 0;
+    dwCtrLength = 0;
```
Thanks, I like your approach to setting up these variables. I think it is more flexible than my original approach.

mback2k commented:
Thanks for looking into this. Are you are right about the first loop. But maybe at least the counter increment function could be made generic, even though openssl.c only uses a fixed length of 16 bytes. I am not sure about the best place to put such a function, maybe misc.c or crypt.c?

madwizard-thomas commented:
I've moved both loops to functions in misc.c. This seemed the most appropriate since crypto.c pretty much just contains the high level crypto method abstractions and not a generic collection of crypto functions.
I've also fixed the code style issue.

mback2k commented on src/wincng.h:
```diff
@@ -299,21 +302,20 @@ struct _libssh2_wincng_cipher_ctx {
     unsigned long dwKeyLength;
-    unsigned long dwUseIV;
+    int useIV;
+    int ctrMode;
```
Thanks for your changes. What do you think about using bool instead of int? I know that bool usually maps to int on most plattforms/toolchains, but the compiler could optimize it since only a boolean value is stored.

madwizard-thomas replied:
Bool would be fine by me, I wasn't sure if it was compatible with the C specs libssh2 adheres to, I didn't see boolean being used anywhere.

mback2k replied:
@bagder What do you think about using 'bool'?

bagder replied:
I actually like using bool, but as we have no code already using it within libssh2 and it isn't a native type in C89 it needs to be used with care if we're going to introduce it now. So, the easy route is to add a comment about it but avoid it for now until we have someone introduce it properly!

mback2k replied:
Okay, thanks for your comment. Sounds like a good plan to me.

madwizard-thomas commented on src/misc.c:
```diff
@@ -643,3 +643,30 @@ void *_libssh2_calloc(LIBSSH2_SESSION* session, size_t size)
+    if (length == 0)
+        return;
+    size_t i = (length - 1);
+    while (ctr[i]++ == 0xFF) {
```
Actually this loop should probably not break early depending on cryptographic data in the counter bytes. It should run in constant time to prevent timing attacks.

mback2k replied:
That's right, but we have to find a way to do something that cannot be optimized away by compilers inside the loop instead of running no-ops.

mback2k commented on src/misc.c:
```diff
@@ -643,3 +643,30 @@ void *_libssh2_calloc(LIBSSH2_SESSION* session, size_t size)
+{
+    if (length == 0)
+        return;
+    size_t i = (length - 1);
```
This line causes issues with some compilers, because size_t i is declared after code. Please move size_t i, to the top of the function. E.g. https://ci.appveyor.com/project/alamaison/libssh2-nkc47/build/168/job/4sapbrepqyhvx2y3

madwizard-thomas commented:
I've implemented the AES CTR code in a time constant manner and added a note about the boolean type. I think everything is fixed now?

mback2k commented:
@madwizard-thomas Thanks a lot for your contribution. I just reordered and reformatted the commits a little bit and just pushed your changes to the master branch on this repository. Consider this pull request merged.
